File: src/gigaphone/adapters/registry.py
"""Backend-adapter registry + selection (DESIGN §9).

Selection: OTel already present → OTel adapter; native SDK present → that adapter; else the
customer's chosen platform. v1 ships the generic OTel default (native adapters land in M6).
"""

from __future__ import annotations

from gigaphone.adapters.backend.braintrust import BraintrustAdapter
from gigaphone.adapters.backend.langsmith import LangSmithAdapter
from gigaphone.adapters.backend.logfire import LogfireAdapter
from gigaphone.adapters.backend.otel import OtelAdapter
from gigaphone.adapters.backend.phoenix import PhoenixAdapter
from gigaphone.interfaces.backend_adapter import BackendAdapter

_BACKENDS: dict[str, BackendAdapter] = {
    "otel": OtelAdapter(),
    "braintrust": BraintrustAdapter(),
    "langsmith": LangSmithAdapter(),
    "logfire": LogfireAdapter(),
    "phoenix": PhoenixAdapter(),
}


def backend_by_id(backend_id: str) -> BackendAdapter | None:
    return _BACKENDS.get(backend_id)


def select_backend(repo: str, preferred: str | None = None) -> BackendAdapter:
    if preferred and preferred in _BACKENDS:
        return _BACKENDS[preferred]
    # Native SDK present → that adapter; else the generic OTel tier (DESIGN §9).
    for native in ("braintrust", "langsmith", "logfire", "phoenix"):
        if _BACKENDS[native].detect_presence(repo):
            return _BACKENDS[native]
    return _BACKENDS["otel"]

File: src/adapters/backend/registry.ts
/**
 * Backend-adapter registry + selection (DESIGN §9).
 *
 * Selection: native SDK present → that adapter; else the generic OTel default tier.
 */

import type { BackendAdapter } from "../../interfaces/backendAdapter.js";
import { BraintrustAdapter } from "./braintrust.js";
import { LangSmithAdapter } from "./langsmith.js";
import { LogfireAdapter } from "./logfire.js";
import { OtelAdapter } from "./otel/adapter.js";
import { PhoenixAdapter } from "./phoenix.js";

const _BACKENDS: Record<string, BackendAdapter> = {
  otel: new OtelAdapter(),
  braintrust: new BraintrustAdapter(),
  langsmith: new LangSmithAdapter(),
  logfire: new LogfireAdapter(),
  phoenix: new PhoenixAdapter(),
};

export function backendById(id: string): BackendAdapter | undefined {
  return _BACKENDS[id];
}

export function selectBackend(repo: string, preferred?: string): BackendAdapter {
  if (preferred && preferred in _BACKENDS) return _BACKENDS[preferred]!;
  // Native SDK present → that adapter; else the generic OTel tier (DESIGN §9).
  for (const native of ["braintrust", "langsmith", "logfire", "phoenix"]) {
    if (_BACKENDS[native]!.detectPresence(repo)) return _BACKENDS[native]!;
  }
  return _BACKENDS.otel!;
}
